File: samples/grpc/iot-service-scala/src/main/scala/iot/registration/RegistrationServiceImpl.scala
package iot.registration

import java.util.concurrent.TimeoutException

import scala.concurrent.Future

import akka.Done
import akka.actor.typed.ActorSystem
import akka.cluster.sharding.typed.scaladsl.ClusterSharding
import akka.grpc.GrpcServiceException
import akka.util.Timeout
import com.google.protobuf.empty.Empty
import io.grpc.Status
import org.slf4j.LoggerFactory

class RegistrationServiceImpl(system: ActorSystem[_])
    extends proto.RegistrationService {

  import system.executionContext

  private val logger = LoggerFactory.getLogger(getClass)

  implicit private val timeout: Timeout =
    Timeout.create(
      system.settings.config.getDuration("iot-service.ask-timeout"))

  private val sharding = ClusterSharding(system)

  override def register(in: proto.RegisterRequest): Future[Empty] = {
    logger.info("register sensor {}", in.sensorEntityId)
    val entityRef = sharding.entityRefFor(Registration.EntityKey, in.sensorEntityId)
    val reply: Future[Done] =
      entityRef.askWithStatus(
        Registration.Register(Registration.SecretDataValue(in.secret), _))
    val response = reply.map(_ => Empty.defaultInstance)
    convertError(response)
  }

  private def convertError[T](response: Future[T]): Future[T] = {
    response.recoverWith {
      case _: TimeoutException =>
        Future.failed(
          new GrpcServiceException(
            Status.UNAVAILABLE.withDescription("Operation timed out")))
      case exc =>
        Future.failed(
          new GrpcServiceException(
            Status.INVALID_ARGUMENT.withDescription(exc.getMessage)))
    }
  }

}

File: samples/grpc/iot-service-java/src/main/java/iot/registration/RegistrationServiceImpl.java
package iot.registration;

import akka.Done;
import akka.actor.typed.ActorSystem;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.EntityRef;
import akka.grpc.GrpcServiceException;
import io.grpc.Status;
import iot.registration.proto.Empty;
import iot.registration.proto.RegisterRequest;
import iot.registration.proto.RegistrationService;
import java.time.Duration;
import java.util.concurrent.CompletionStage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RegistrationServiceImpl implements RegistrationService {
  private final Logger logger = LoggerFactory.getLogger(RegistrationServiceImpl.class);
  private final ActorSystem<?> system;
  private final Duration timeout;
  private final ClusterSharding sharding;

  public RegistrationServiceImpl(ActorSystem<?> system) {
    this.system = system;
    this.timeout = system.settings().config().getDuration("iot-service.ask-timeout");
    this.sharding = ClusterSharding.get(system);
  }

  @Override
  public CompletionStage<Empty> register(RegisterRequest in) {
    logger.info("register sensor {}", in.getSensorEntityId());
    EntityRef<Registration.Command> entityRef =
        sharding.entityRefFor(Registration.ENTITY_KEY, in.getSensorEntityId());
    CompletionStage<Done> response =
        entityRef.askWithStatus(
            replyTo ->
                new Registration.Register(
                    new Registration.SecretDataValue(in.getSecret()), replyTo),
            timeout);
    return response.handle(
        (done, error) -> {
          if (error != null) {
            throw new GrpcServiceException(Status.INTERNAL.withDescription(error.getMessage()));
          } else {
            return Empty.getDefaultInstance();
          }
        });
  }
}
